#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
å·¥ä¼¤æ¡ˆä»¶ç®¡ç†ç³»ç»Ÿ - ä¸»ç¨‹åº
"""
import os
import sys
from datetime import datetime
from PyQt5.QtWidgets import QApplication, QMainWindow
from PyQt5.uic import loadUi
from PyQt5.QtCore import Qt, QSettings
from openpyxl import load_workbook
from config_manager import ConfigManager


class MainWindow(QMainWindow):

    def __init__(self):
        super().__init__()

        # 1. åŠ è½½ç•Œé¢
        loadUi("main_window.ui", self)
        self.setWindowTitle("å·¥ä¼¤æ¡ˆä»¶ç®¡ç†ç³»ç»Ÿ")

        # 2. åˆå§‹åŒ–é…ç½®ç®¡ç†å™¨
        self.config = ConfigManager()

        # 3. åŠ è½½Excelæ•°æ®åˆ°ComboBox
        self.load_excel_to_combobox()

        # 4. åŠ è½½ä¿å­˜çš„é…ç½®
        self.load_config()

        # 5. è¿æ¥ä¿¡å·
        self.checkBox_remember.stateChanged.connect(self.on_remember_changed)
        self.btn_generate_record.clicked.connect(self.on_generate_record)

        # 6. æ ¹æ®è®°ä½çŠ¶æ€æ›´æ–°ç•Œé¢
        self.update_ui()

    def determine_word_template(self, person_type, case_type, regulation_key):
        """æ ¹æ®æ¡ä»¶ç¡®å®šWordæ¨¡æ¿æ–‡ä»¶è·¯å¾„"""

        # 1. æ£€æŸ¥æ˜¯å¦æ˜¯æœ¬äºº + æ™®é€šæ¡ˆä»¶ï¼ˆæœ€ç®€å•çš„æƒ…å†µï¼‰
        if person_type == "æœ¬äºº" and case_type == "æ™®é€šæ¡ˆä»¶":
            # å…ˆæ£€æŸ¥æ˜¯å¦æœ‰å¯¹åº”çš„æ¨¡æ¿æ–‡ä»¶
            template_path = "templates/æœ¬äººæ™®é€šæ¡ˆä»¶æ¨¡æ¿.docx"
            if os.path.exists(template_path):
                return template_path
            else:
                # å¦‚æœæ–‡ä»¶ä¸å­˜åœ¨ï¼Œåˆ›å»ºä¸€ä¸ªç®€å•çš„æç¤º
                self.statusBar().showMessage("æ¨¡æ¿æ–‡ä»¶ä¸å­˜åœ¨: " + template_path, 3000)
                return None

        # 2. å…¶ä»–æƒ…å†µæš‚æ—¶è¿”å›é»˜è®¤æ¨¡æ¿
        else:
            # å¯ä»¥åœ¨è¿™é‡Œæ·»åŠ æ›´å¤šçš„æ¨¡æ¿åˆ¤æ–­é€»è¾‘
            default_template = "templates/é€šç”¨æ¨¡æ¿.docx"
            if os.path.exists(default_template):
                self.statusBar().showMessage(f"ä½¿ç”¨é€šç”¨æ¨¡æ¿: {person_type}+{case_type}", 3000)
                return default_template
            else:
                self.statusBar().showMessage("æ¨¡æ¿æ–‡ä»¶ä¸å­˜åœ¨ï¼Œè¯·æ£€æŸ¥templatesç›®å½•", 3000)
                return None

    def load_excel_to_combobox(self):
        """ä»Excelæ–‡ä»¶åŠ è½½æ•°æ®åˆ°ComboBox"""
        # è·å–å½“å‰ç›®å½•
        current_dir = os.path.dirname(os.path.abspath(__file__))

        # è¯»å–ç”¨äººå•ä½Excel
        try:
            employer_file = os.path.join(current_dir, "ç”¨äººå•ä½åç§°æ±‡æ€».xlsx")
            if os.path.exists(employer_file):
                wb = load_workbook(employer_file)
                ws = wb.active
                # è¯»å–ç¬¬ä¸€åˆ—æ‰€æœ‰æœ‰æ•°æ®çš„å•å…ƒæ ¼
                for row in ws.iter_rows(min_row=1, max_col=1, values_only=True):
                    if row[0]:  # æ£€æŸ¥å•å…ƒæ ¼æ˜¯å¦ä¸ºç©º
                        self.comboBox_employer.addItem(str(row[0]))
        except Exception as e:
            print(f"è¯»å–ç”¨äººå•ä½Excelå¤±è´¥: {e}")

        # è¯»å–ç”¨å·¥å•ä½Excel
        try:
            work_unit_file = os.path.join(current_dir, "ç”¨å·¥å•ä½åç§°æ±‡æ€».xlsx")
            if os.path.exists(work_unit_file):
                wb = load_workbook(work_unit_file)
                ws = wb.active
                for row in ws.iter_rows(min_row=1, max_col=1, values_only=True):
                    if row[0]:
                        self.comboBox_work_unit.addItem(str(row[0]))
        except Exception as e:
            print(f"è¯»å–ç”¨å·¥å•ä½Excelå¤±è´¥: {e}")

        # è¯»å–å·¥ä½œåœºæ‰€Excel
        try:
            workplace_file = os.path.join(current_dir, "å·¥ä½œåœºæ‰€åç§°æ±‡æ€».xlsx")
            if os.path.exists(workplace_file):
                wb = load_workbook(workplace_file)
                ws = wb.active
                for row in ws.iter_rows(min_row=1, max_col=1, values_only=True):
                    if row[0]:
                        self.comboBox_workplace.addItem(str(row[0]))
        except Exception as e:
            print(f"è¯»å–å·¥ä½œåœºæ‰€Excelå¤±è´¥: {e}")

    def load_config(self):
        """åŠ è½½é…ç½®åˆ°ç•Œé¢"""
        config = self.config.load_config()

        # è®¾ç½®æ§ä»¶å†…å®¹
        self.lineEdit_operator.setText(config['operator'])
        self.lineEdit_api_url.setText(config['api_url'])
        self.lineEdit_api_key.setText(config['api_key'])
        self.checkBox_remember.setChecked(config['remember'])

    def update_ui(self):
        """æ›´æ–°ç•Œé¢çŠ¶æ€"""
        remember = self.checkBox_remember.isChecked()

        # è®¾ç½®è¾“å…¥æ¡†æ˜¯å¦å¯ç¼–è¾‘
        self.lineEdit_operator.setEnabled(not remember)
        self.lineEdit_api_url.setEnabled(not remember)
        self.lineEdit_api_key.setEnabled(not remember)

        # è®¾ç½®æ ·å¼ï¼ˆç°è‰²èƒŒæ™¯è¡¨ç¤ºä¸å¯ç¼–è¾‘ï¼‰
        if remember:
            style = "background-color: #f0f0f0; color: #666;"
            self.statusBar().showMessage("é…ç½®å·²è®°ä½ï¼Œå–æ¶ˆå‹¾é€‰å¯ä¿®æ”¹", 2000)
        else:
            style = ""

        self.lineEdit_operator.setStyleSheet(f"QLineEdit {{ {style} }}")
        self.lineEdit_api_url.setStyleSheet(f"QLineEdit {{ {style} }}")
        self.lineEdit_api_key.setStyleSheet(f"QLineEdit {{ {style} }}")

    def on_remember_changed(self):
        """è®°ä½æˆ‘å¤é€‰æ¡†çŠ¶æ€å˜åŒ–"""
        remember = self.checkBox_remember.isChecked()

        if remember:
            # ä¿å­˜å½“å‰é…ç½®
            operator = self.lineEdit_operator.text().strip()
            api_url = self.lineEdit_api_url.text().strip()
            api_key = self.lineEdit_api_key.text().strip()

            self.config.save_config(operator, api_url, api_key, True)
            self.statusBar().showMessage("é…ç½®å·²ä¿å­˜", 1500)
        else:
            # æ¸…é™¤é…ç½®
            self.config.clear_config()
            self.statusBar().showMessage("é…ç½®å·²æ¸…é™¤", 1500)

        # æ›´æ–°ç•Œé¢
        self.update_ui()

    def check_case_type(self):
        """æ£€æŸ¥æ¡ˆä»¶ç±»å‹"""
        is_personal = self.checkBox_personal.isChecked()
        is_death = self.checkBox_death.isChecked()

        if is_personal and is_death:
            return "ä¸ªäººç”³è¯·æ­»äº¡æ¡ˆä»¶"
        elif is_personal:
            return "ä¸ªäººæ¡ˆä»¶"
        elif is_death:
            return "æ­»äº¡æ¡ˆä»¶"
        else:
            return "æ™®é€šæ¡ˆä»¶"

    def check_person_type(self):
        """æ£€æŸ¥äººå‘˜ç±»å‹"""
        if self.radio_self.isChecked():
            return "æœ¬äºº"
        elif self.radio_witness.isChecked():
            return "è¯äºº"
        elif self.radio_legal_entity.isChecked():
            return "æ³•äºº"

    def calculate_id_info(self, id_card):
        """æ ¹æ®èº«ä»½è¯å·è®¡ç®—å¹´é¾„å’Œæ€§åˆ«"""
        if len(id_card) != 18:
            return id_card, None, None

        # æå–å‡ºç”Ÿå¹´æœˆæ—¥
        birth_year = int(id_card[6:10])
        birth_month = int(id_card[10:12])
        birth_day = int(id_card[12:14])

        # è®¡ç®—å¹´é¾„
        current_year = datetime.now().year
        current_month = datetime.now().month
        current_day = datetime.now().day

        age = current_year - birth_year
        if current_month < birth_month or (current_month == birth_month and current_day < birth_day):
            age -= 1

        # è®¡ç®—æ€§åˆ«ï¼ˆç¬¬17ä½ï¼Œå¥‡æ•°ä¸ºç”·ï¼Œå¶æ•°ä¸ºå¥³ï¼‰
        gender_num = int(id_card[16])
        gender = "ç”·" if gender_num % 2 == 1 else "å¥³"

        return id_card, age, gender

    def on_generate_record(self):
        """ç”Ÿæˆç¬”å½•æŒ‰é’®ç‚¹å‡»äº‹ä»¶"""
        print("ğŸ“ ç”Ÿæˆç¬”å½•æŒ‰é’®è¢«ç‚¹å‡»")

        # æ£€æŸ¥æ¡ˆä»¶ç±»å‹
        case_type = self.check_case_type()
        print(f"æ¡ˆä»¶ç±»å‹: {case_type}")

        # æ£€æŸ¥äººå‘˜ç±»å‹
        person_type = self.check_person_type()
        print(f"äººå‘˜ç±»å‹: {person_type}")

        # å¦‚æœæ˜¯æœ¬äººç±»å‹ï¼Œæ£€æŸ¥å§“åå¹¶å¤åˆ¶åˆ°å—ä¼¤èŒå·¥
        if person_type == "æœ¬äºº":
            name = self.lineEdit_name.text().strip()
            print(f"æœ¬äººå§“å: '{name}'")

            if name:
                self.lineEdit_injured_worker.setText(name)
                print("âœ… å§“åå·²å¤åˆ¶åˆ°å—ä¼¤èŒå·¥")
            else:
                self.statusBar().showMessage("æœ¬äººä¿¡æ¯æœªå¡«å†™", 3000)
                print("é”™è¯¯ï¼šæœ¬äººä¿¡æ¯æœªå¡«å†™")
                return

        # å¤„ç†èº«ä»½è¯ä¿¡æ¯
        id_card = self.lineEdit_id_card.text().strip()
        if id_card:
            id_card, age, gender = self.calculate_id_info(id_card)
            if age and gender:
                self.lineEdit_age.setText(str(age))
                self.comboBox_gender.setCurrentText(gender)

        # è·å–å…¶ä»–åŸºæœ¬ä¿¡æ¯
        id_address = self.lineEdit_id_address.text().strip()
        current_address = self.lineEdit_current_address.text().strip()
        phone = self.lineEdit_phone.text().strip()
        position = self.lineEdit_position.text().strip()

        # è·å–æ‹Ÿç”¨æ¡ä¾‹
        regulation_index = self.comboBox_regulations.currentIndex()
        regulation_mapping = {
            0: "ç¬¬åå››æ¡ç¬¬ä¸€æ¬¾ç¬¬ä¸€é¡¹",
            1: "ç¬¬åå››æ¡ç¬¬ä¸€æ¬¾ç¬¬äºŒé¡¹",
            2: "ç¬¬åå››æ¡ç¬¬ä¸€æ¬¾ç¬¬ä¸‰é¡¹",
            3: "ç¬¬åå››æ¡ç¬¬ä¸€æ¬¾ç¬¬å››é¡¹",
            4: "ç¬¬åå››æ¡ç¬¬ä¸€æ¬¾ç¬¬äº”é¡¹",
            5: "ç¬¬åå››æ¡ç¬¬ä¸€æ¬¾ç¬¬å…­é¡¹",
            6: "ç¬¬åäº”æ¡ç¬¬ä¸€æ¬¾ç¬¬ä¸€é¡¹"
        }
        regulation_key = regulation_mapping.get(regulation_index, "æœªçŸ¥æ¡ä¾‹")

        # è·å–å•ä½ä¿¡æ¯
        employer = self.comboBox_employer.currentText().strip()
        work_unit = self.comboBox_work_unit.currentText().strip()
        workplace = self.comboBox_workplace.currentText().strip()

        # ====== æ‰“å¼€Wordæ¨¡æ¿ ======
        print(f"å½“å‰ç›®å½•: {os.path.dirname(__file__)}")

        # å…ˆæµ‹è¯•ç›´æ¥æ‰“å¼€
        template_path = os.path.join(os.path.dirname(__file__), "templates", "æœ¬äººæ™®é€šæ¡ˆä»¶æ¨¡æ¿.docx")
        print(f"æ¨¡æ¿è·¯å¾„: {template_path}")
        print(f"æ¨¡æ¿å­˜åœ¨: {os.path.exists(template_path)}")

        if os.path.exists(template_path):
            print(f"âœ… æ‰¾åˆ°æ¨¡æ¿ï¼Œå‡†å¤‡æ‰“å¼€Wordæ–‡ä»¶")

            # ç®€å•æµ‹è¯•ï¼šç›´æ¥æ‰“å¼€
            os.startfile(template_path)  # Windowsç›´æ¥æ‰“å¼€

            # æˆ–è€…ä½¿ç”¨ä½ çš„å®Œæ•´æ–¹æ³•
            # self.open_word_template(template_path, {
            #     'æ¡ˆä»¶ç±»å‹': case_type,
            #     'äººå‘˜ç±»å‹': person_type,
            #     'æ¡ä¾‹': regulation_key,
            #     'å§“å': self.lineEdit_name.text().strip(),
            #     'å¹´é¾„': age if 'age' in locals() and age else '',
            #     'æ€§åˆ«': gender if 'gender' in locals() and gender else '',
            #     'èº«ä»½è¯å·': id_card if id_card else '',
            #     'èº«ä»½è¯åœ°å€': id_address,
            #     'ç°ä½å€': current_address,
            #     'ç”µè¯': phone,
            #     'å²—ä½': position,
            #     'ç”¨äººå•ä½': employer,
            #     'ç”¨å·¥å•ä½': work_unit,
            #     'å·¥ä½œåœºæ‰€': workplace
            # })

            self.statusBar().showMessage("å·²æ‰“å¼€Wordæ–‡ä»¶", 3000)
        else:
            print(f"âŒ æ¨¡æ¿ä¸å­˜åœ¨")
            # åˆ—å‡ºtemplatesç›®å½•å†…å®¹
            templates_dir = os.path.join(os.path.dirname(__file__), "templates")
            if os.path.exists(templates_dir):
                files = os.listdir(templates_dir)
                print(f"templatesç›®å½•ä¸­çš„æ–‡ä»¶: {files}")
            else:
                print(f"templatesç›®å½•ä¸å­˜åœ¨")

            self.statusBar().showMessage("æ¨¡æ¿æ–‡ä»¶ä¸å­˜åœ¨", 3000)

        # æ˜¾ç¤ºç»“æœ
        result = f"æ¡ˆä»¶ç±»å‹: {case_type}, äººå‘˜ç±»å‹: {person_type}, æ¡ä¾‹: {regulation_key}"
        print(result)

    def closeEvent(self, event):
        """çª—å£å…³é—­æ—¶æœ€åä¿å­˜ä¸€æ¬¡"""
        if self.checkBox_remember.isChecked():
            operator = self.lineEdit_operator.text().strip()
            api_url = self.lineEdit_api_url.text().strip()
            api_key = self.lineEdit_api_key.text().strip()
            self.config.save_config(operator, api_url, api_key, True)

        # ä¿å­˜çª—å£å¤§å°ä½ç½®
        settings = QSettings("WorkInjuryApp", "Window")
        settings.setValue("geometry", self.saveGeometry())

        event.accept()


def main():
    app = QApplication(sys.argv)

    # è®¾ç½®åº”ç”¨ä¿¡æ¯
    app.setApplicationName("å·¥ä¼¤æ¡ˆä»¶ç®¡ç†ç³»ç»Ÿ")
    app.setOrganizationName("WorkInjuryApp")

    # åˆ›å»ºå¹¶æ˜¾ç¤ºçª—å£
    window = MainWindow()
    window.show()

    sys.exit(app.exec_())


if __name__ == "__main__":
    main()